libraries/cbor/src/reader.rs

```diff
@@ -37,11 +37,16 @@ pub enum DecoderError {
     OutOfRangeIntegerValue,
 }
 
-/// Deserialize CBOR binary data to produce a single [`Value`], expecting that there
-/// is no additional data.
+/// Deserialize CBOR binary data to produce a single [`Value`], expecting that there is no additional data.
 pub fn read(encoded_cbor: &[u8]) -> Result<Value, DecoderError> {
+    read_nested(encoded_cbor, Reader::MAX_NESTING_DEPTH)
+}
+
+/// Deserialize CBOR binary data to produce a single [`Value`], expecting that there is no additional data.  Nested
+/// structures up to the provided `max_nest` are supported.
+pub fn read_nested(encoded_cbor: &[u8], max_nest: i8) -> Result<Value, DecoderError> {
     let mut reader = Reader::new(encoded_cbor);
-    let value = reader.decode_complete_data_item(Reader::MAX_NESTING_DEPTH)?;
```

Comment:
Since we are splitting, `MAX_NESTING_DEPTH` in `Reader` and `Writer` should probably be a constant in `ctap2`, as it is not part of CBOR. It means that the current `read` with one parameter should be moved to the application. If we do that, one would arguably also want it to return a `Ctap2StatusCode` as error, and then do some refactoring to always just `?`. Let me do that in a next PR, so we just have to agree on the interface we want for the library.

I don't think there is a general default for nesting depth in CBOR. Maybe we still want a default of infinity or something very big for convenience. So IMO it could be:
- `read_nested` and `read` with a big default max or the depth becomes an `Option`
- a variety of `read_VARIANT` functions for each variant of CBOR that we want to support, if such a list exists, like `read_ctap`, `read_canonical`, `read_crazy_variant`

This comment applies to `writer` as well.

Reply:
So would it make sense to have:
 - `read(encoded_cbor: &[u8]) -> Result<Value, DecoderError>` use an infinite depth (and therefore document that it can never return `Err(DecoderError::TooMuchNesting)`), by passing `None` as the 2nd argument to &hellip;
 - `read_nested(encoded_cbor: &[u8], max_nest: Option<i8>) -> Result<Value, DecoderError>`.

Then the CTAP-specific stuff (namely the `MAX_NESTING_DEPTH`) can live in the ctap-specific code, and use `read_nested(data, Some(MAX_NESTING_DEPTH))`.

(To make `writer` work similarly will probably need the next PR I've got pending to be folded in, which changes `write` to return a `Result`)

WDYT?

Reply:
Sounds good to me. Adding the `Option` shouldn't be more than a bunch of `map` and `if let`.

Since we don't want to break / change OpenSK in transition, maybe `read` passes `Some(4)` in the second argument in this PR, and I'll then port it over? So you don't have to dig into more OpenSK code than necessary? (Of course I don't mind if you do :)

Reply:
I've re-jigged this PR as suggested, and tried to adapt the OpenSK code that uses cbor at the same time.  If the CI actions pass, is that likely to cover everything that's needed?

Reply:
Yeah, should be okay. For convenience, there is also `./run_desktop_tests.sh` similar to the GitHub actions. It is currently broken on develop with the `std` removal in one of the last PRs. I'll submit a fix soon.

Reply:
Ah, sorry, I didn't know about that one.

Reply:
No problem, fixed.